didcomm/protocols/forward/forward.py

```diff
@@ -24,37 +80,161 @@ class ForwardResult:
+        #      (e.g. first one may use alternative endpoint but second - URI)
+
+        # resolve until final URI is reached
+        mediator_did_service = await find_did_service(
+           resolvers_config, mediator_did
+        )
+        if mediator_did_service is None:
+            raise DIDServiceNotFoundError(
+                f"mediator '{mediator_did}' service doc not found"
+            )
+
+        service_uri = mediator_did_service.service_endpoint
+        res.append(mediator_did_service)
+
+    return res
+
+
 async def wrap_in_forward(
     resolvers_config: ResolversConfig,
     packed_msg: Union[JSON_OBJ, JSON],
-    routing_key_ids: List[DID_OR_DID_URL],
-    forward_headers: Optional[List[Header]] = None,
-) -> JSON:
+    to: DID_OR_DID_URL,
+    routing_keys: List[DID_OR_DID_URL],
+    enc_alg_anon: Optional[AnonCryptAlg] = DEF_ENC_ALG_ANON,
+    headers: Optional[List[Header]] = None,
+    didcomm_id_generator: Optional[DIDCommGeneratorType] = None,
+) -> Optional[ForwardPackResult]:
     """
+    Resolves recipient DID DOC Service and Builds Forward envelops if needed.
+
     Wraps the given packed DIDComm message in Forward messages for every routing key.
 
     :param resolvers_config: secrets and DIDDoc resolvers
     :param packed_msg: the message to be wrapped in Forward messages
-    :param routing_key: a list of routing keys
-    :param forward_headers: optional headers for Forward message
+    :param to: recipient's DID (DID URL)
+    :param routing_keys: list of routing keys
+    :param enc_alg_anon (AnonCryptAlg): The encryption algorithm to be used for
+        anonymous encryption (anon_crypt).
+    :param headers: optional headers for Forward message
+    :param didcomm_id_generator (Callable): optional callable to use
+        for forward messages ``id`` generation, ``didcomm_id_generator_default``
+        is used by default
 
     :raises DIDDocNotResolvedError: If a DID can not be resolved to a DID Doc.
     :raises DIDUrlNotFoundError: If a DID URL (for example a key ID) is not found within a DID Doc
     :raises SecretNotFoundError: If there is no secret for the given DID or DID URL (key ID)
 
     :return: a top-level packed Forward message as JSON string
     """
-    return json.dumps(
-        ForwardMessage(
-            body=ForwardBody(next=""),
-            id="",
-            type="https://didcomm.org/routing/2.0/forward",
-        ).as_dict()
+
+    headers = headers or {}  # TODO headers validation against ForwardMessage
+
+    # means forward protocol is not needed
+    if not routing_keys:
+        logger.debug("No routing keys found: skipping forward wrapping")
+        return None
+
+    # wrap forward msgs in reversed order so the message to final
+    # recipient 'to' will be the innermost one
+    for _to, _next in zip(
+        routing_keys[::-1], (routing_keys[1:] + [to])[::-1]
+    ):
+        fwd_attach = Attachment(
+            data=AttachmentDataJson(packed_msg)
```

Comment:
`AttachmentDataJson` expect a `JSON_VALUE` here (for example a Dict), while `wrap_in_forward` declares `packed_msg` as a `Union[JSON_OBJ, JSON]`. I think it will not work if we pass `JSON` (that is just str) there. 
So,
1) We need to fix it. Can we consider accepting a `JSON_VALUE` (dict) only as a `packed_msg`? It will probably simplify everything...
2) In `pack_encrypted`: we do dict_to_json conversion before forward, then pass it (JSON) to forward, get dict, and call dict_to_json again. Can we do dict_to_json only once (at the very end), and pass dict to wrap_in_forward?

Reply:
updated